### PR TITLE
prometheusexporter: Automatic normalization of metrics from Otel to Prometheus naming convention

### PR DESCRIPTION
**Pull request by bertysentry**
**Description:** In **prometheusexporter** and **prometheusremotewriteexporter**, automatic normalization of metrics from [OpenTelemetry semantic convention](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md) to [Prometheus naming convention](https://prometheus.io/docs/practices/naming/)

The renaming is in **alpha** stage and thus disabled by default (behind a feature gate). Enable with `--feature-gates=pkg.translator.prometheus.NormalizeName` option.

Shared code between **prometheusexporter** and **prometheusremotewriteexporter** has been moved to a new module: `pkg/translator/prometheus`. This module provides 2 main functions:

* `func NormalizeLabel(label string) string` for labels normalization
* `func BuildPromCompliantName(metric pmetric.Metric, namespace string) string` for metric names normalization

All normalization operations are detailed in `pkg/translator/prometheus/README.md`.

> **Note**
> It is recommended to import `pkg/translator/prometheus` as `prometheustranslator` to avoid conflict with the "real" Prometheus module, from Prometheus.

**Link to tracking Issue:** #8950

**Testing:** See unit tests in the `pkg/translator/prometheus` module

**Documentation:** Updated README.md in **prometheusexporter** and **prometheusremotewriteexporter**

### CODE REVIEW

**Comment by Aneurysm9** [comment]
This looks reasonable to me, though I'm still not clear on whether `_total` should be appended to counter names or not.  I think some prometheus client libraries do that and some don't.

What happens if metric names already contain a suffix or go through multiple round trips?  Will the suffix be omitted if it already exists?

**Comment by bertysentry** [comment]
@Aneurysm9 Thank you for the comment!

`_total` must be appended to *Counter* metrics as per [Prometheus naming standard](https://prometheus.io/docs/practices/naming/).

Good point about round-trip conversion: units **must not** be appended when **already** present in the metric name (this goes for units, `_total` and `_ratio` suffixes).

Also, the issue #8950 mentions the need for the Prometheus receiver to perform the reverse "conversion", like in the example below:

`system_io_bytes_total` ==> `system.io` *Counter* with Units `By`

This should be covered in a separate PR.

**Review by dashpole — COMMENTED** [review]
Whoops, forgot to click send...

File: exporter/prometheusexporter/README.md
Comment (bertysentry) [reply exporter/prometheusexporter/README.md:25]: I understand your concerns.

The problem is that processors run **before** exporters in the pipeline. The idea here is to be able to override this automatic conversion with a manually-defined mapping, for a few metrics. This is just a workaround mechanism to allow the user to cover edge cases (when one Otel metric is badly defined, etc.) 

Instead of `rename`, we could have a `disable_automatic_renaming` option where the user could list the metrics that should be left unchanged (the necessary renaming would have been taken care of in **metrictransformprocessor**). But it's making things more complicated for the end user, and code wouldn't be much simpler.

File: exporter/prometheusexporter/README.md
Comment (bertysentry) [reply exporter/prometheusexporter/README.md:66]: Totally. `_ratio` will be appended only for metrics whose unit is `1`, as per OpenTelemetry's metric semantic convention. `_ratio` will not be appended to unitless metrics.

File: exporter/prometheusexporter/README.md
Comment (bertysentry) [reply exporter/prometheusexporter/README.md:65]: That's not what I understand from OpenMetrics:

> If non-base units can not be avoided and conversion is not feasible, the actual unit should still be included in the metric name for clarity.

So non-standard unit should be appended to the metric name (if not already present in the metric name).

File: exporter/prometheusexporter/README.md
Comment (bertysentry) [reply exporter/prometheusexporter/README.md:22]: I agree with you, I'm just concerned with the potential problems this may cause. How do you see this feature-gate working? I'm confused on how that would work.

Also, important question: Are we trying to comply with [Prometheus' metric naming convention](https://prometheus.io/docs/practices/naming/) or with [OpenMetrics](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md)? OpenMetrics has more guidelines (like appending `_sum` for *Summary* metrics, `_gcount` and `_gsum` for *GaugeHistogram*, etc.)

File: exporter/prometheusexporter/README.md
Comment (pmm-sumo) [reply exporter/prometheusexporter/README.md:25]: > The problem is that processors run before exporters in the pipeline.

@bertysentry I am not sure if it add that much more complexity. Just one more processor to refer to. And most users probably will use single exporter anyway.

If you want to have e.g. two exporters and make changes only for one of them, then actually [two pipelines referring the same receiver](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/design.md#receivers) can be created (so e.g. one with metricstransformprocessor and the other without it). So we currently have the flexibility to do it, perhaps it needs to be documented better

Comment (bertysentry) [reply exporter/prometheusexporter/README.md:25]: @pmm-sumo Ha! I totally missed the ability to define several pipelines! 😅

But we still need an option to disable the automatic conversion from Otel to Prometheus/OpenMetrics standard for certain metrics, right?

Example: 

We have a bunch of metrics that are properly converted from Otel to Prometheus standard, except for this one:
*Gauge* `myapp.filesystem.usage_megabytes` with unit `{MBy}` (obviously broken, but it happens)

Automatic conversion would produce `myapp_filesystem_usage_megabytes_MBy` (`MBy` was appended, even though `_megabytes` is already mentioned).

User wants `myapp_filesystem_usage_megabytes` (no `_MBy`).

User will not be able to **metrictransformprocessor** because there is no renaming to be done, and this processor cannot remove the `MBy` unit from metric.

I'm fine simply removing the `rename` map option and even the ability to cover the above example (the code will be just simpler). I just want to be sure I'm not going to be asked to put it back later, ha ha 😉

Comment (bertysentry) [reply exporter/prometheusexporter/README.md:25]: @pmm-sumo @dashpole Should we remove the `namespace` option from the Prometheus exporter? It could be handled with a processor as well. (in a separate PR of course!)

File: exporter/prometheusexporter/README.md
Comment (dashpole) [reply exporter/prometheusexporter/README.md:22]: See [this](https://github.com/open-telemetry/opentelemetry-collector/tree/e8dd9536f5e53228606c325f12d4bfc4b4293acc/service/featuregate#collector-feature-gates) for the overview of feature-gates.  They are a CLI flag which can enable/disable functionality, and it enables progression for a feature from Alpha (off by default; can be enabled) to Beta (on by default, can be disabled) to GA (on by default, cannot be disabled).  Feature-gates should be used in cases where long-term configuration is not needed, but a more "careful" transition from one state to another is needed.

We start with the feature-gate disabled by default, while we validate that this is a change we _actually_ want to make, and make sure other pieces of functionality (e.g. changes in the prometheus receiver) are in-place.  Then we graduate the feature-gate(s) to beta, and get user-feedback.  If users encounter issues, they can always disable the functionality.

Technically, we should be complying with prometheus' conventions, since [we don't](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1274b14412ede4725184ad17dc8f90f0473f79ea/exporter/prometheusexporter/prometheus.go#L58) [enable openmetrics](https://github.com/prometheus/client_golang/blob/1f81b3e9130fe21c55b7db82ecb2d61477358d61/prometheus/promhttp/http.go#L156) on the prometheus handler.  However, note that suffixing based on openmetrics vs prometheus text format is handled by the prometheus client ([example where openmetrics appends _total](https://github.com/prometheus/common/blob/main/expfmt/openmetrics_create.go#L90)).  So as long as we use the prometheus client, we adding type suffixes correctly.  Units, on the other hand, we need to do ourselves.

File: exporter/prometheusexporter/README.md
Comment (dashpole) [reply exporter/prometheusexporter/README.md:25]: I would support that.

Comment (dashpole) [reply exporter/prometheusexporter/README.md:25]: > Automatic conversion would produce myapp_filesystem_usage_megabytes_MBy (MBy was appended, even though _megabytes is already mentioned).

I think suffixing is actually the correct behavior, given the metric and unit.  Omitting _MBy would result in the output being invalid openmetrics.  It may look better, but we should ensure our output is valid so it isn't rejected by other ecosystem components, which may rely on that behavior.

> But we still need an option to disable the automatic conversion from Otel to Prometheus/OpenMetrics standard for certain metrics, right?

We are discussing that [below](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/10028#discussion_r873711876), but i'm not convinced that we need to be able to disable conversion to OM standards.

I would support the ability for a processor to modify metric units to potentially help in cases like this.

Comment (bertysentry) [reply exporter/prometheusexporter/README.md:25]: Okay I will remove these options. Thanks!

File: exporter/prometheusexporter/README.md
Comment (bertysentry) [reply exporter/prometheusexporter/README.md:65]: Okay, I think I just understood: for non-standard units, the unit must be appended to the metric name (if not already present) and the `Unit` field emptied so that it doesn't end up in `# UNIT ...` when exported. Am I correct?

Comment (bertysentry) [reply exporter/prometheusexporter/README.md:65]: Okay, to summarize:

| Otel Metric Name | Unit | Prometheus Metric |
|---|---|---|
| `astro.galaxy.distance` | `m` | `astro_galaxy_distance_meter` |
| `astro.galaxy.distance` | `ly` ([c/s notation for light-years](https://ucum.org/ucum.html#const)) | `astro_galaxy_distance_ly` |
| `router.network.udp_traffic` | `{packets}` | `router_network_udp_traffic` |

What about broken cases (receivers are not all equal, and some metrics may not follow Otel's semantic convention?

| Otel Metric Name | Unit | Prometheus Metric |
|---|---|---|
| `router.network.udp_traffic` | `packets` (no brackets) | `router_network_udp_traffic_packets` or `router_network_udp_traffic`? |
| `router.network.bandwidth_utilization` | `percent` | `router_network_bandwidth_utilization_percent`? |
|`router.network.bandwidth_utilization` | `%` | `router_network_bandwidth_utilization`? |

Comment (dashpole) [reply exporter/prometheusexporter/README.md:65]: For broken cases, we should probably assume it is a nonstandard unit we haven't heard of, and treat it the same as you do `ly` above.  For units that include characters that aren't allowed in metric names, we can either drop them entirely, or just drop the invalid characters.  I'm inclined to start by dropping the unit entirely in that case.

**Comment by dashpole** [comment]
readme LGTM

**Comment by bertysentry** [comment]
@Aneurysm9 @dashpole @pmm-sumo Code pushed, according to our discussion.

Next step: update **prometheusremotewriteexporter** the same way.

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [inline exporter/prometheusexporter/metric_rename.go:66]: If we aren't going to do this for all known units, can we omit this for now?

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:23]: The official metric semantic convention all use UCUM, and it's the currently the [recommended format in the specifications](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#instrument-units)

I will add links in the code.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:48]: True, but metrics don't necessarily come from **prometheusreceiver** and we still want to make sure `_total` is removed for a *Counter*, so we can process the unit and then re-add `_total` at the end.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:23]: Will change for `unitMap`.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:66]: Yep, I knew you wouldn't like this 😁

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [reply exporter/prometheusexporter/metric_rename.go:23]: Ah, cool.  I couldn't find that in the spec for some reason.

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [reply exporter/prometheusexporter/metric_rename.go:48]: Are there other formats where we expect `_total` to be a common suffix?

Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:48]: Nope, not that I know of. Also, we're trimming it if and only if the metric is a *Monotonic Sum* (i.e. a *Counter*), in which case we will be appending `_total` after the unit.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:59]: Good point! We discussed that case earlier and said we would drop units with non-alphanumeric chars (like `%`)

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [reply exporter/prometheusexporter/metric_rename.go:48]: I'm fine with trimming _total if there are cases we know of in which metrics commonly have that suffix.  But i'd prefer to omit it from the initial version if we don't know of any cases.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go]: @dashpole Okay, I pushed an update with "improved" support for some special chars. Non-supported chars are now properly dropped (redundant underscores are trimmed, etc.). If you don't like it, we'll remove it 😉

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:52]: It's important to trim `_total` because we're appending it again for *Monotonic Sums*. Otherwise, we may end up going from `network.received_bytes_total` (with unit `By`) to `network_received_bytes_total_bytes_total`.

The same logic applies for other suffixes that we append to the metric name, like the unit: if it's already present in the metric name, we don't append it.

`_total` is a bit special because metrics names are typically structured as `<namespace>.<object>.<metric>[_<unit>][_total]`, so we need to remove it to ensure we check the `[_<unit>]` suffix, and then re-add `_total` for *Monotonic Sums*.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go]: So, here is the list of units currently used in Otel's "official" receivers:

* 1
* %
* By
* By/s
* connections
* GiBy/h
* KBy
* KiBy
* MiBy
* ms
* ns
* requests
* s
* scoreboard
* us
* {alerts}
* {assertions}/s
* {attempts}
* {binds}/s
* {brokers}
* {collections}
* {commands}
* {commits}
* {commits}/s
* {compilations}/s
* {connections}
* {consumers}
* {cursors}
* {databases}
* {documents}
* {documents}/s
* {errors}
* {events}
* {evictions}
* {extents}
* {faults}
* {faults}/s
* {file_descriptors}
* {files}
* {flushes}/s
* {followers}
* {growths}
* {indexes}
* {inodes}
* {items}
* {locks}
* {lookups}
* {members}
* {messages}
* {nodes}
* {notifications}
* {objects}
* {objects}/s
* {operation}
* {operations}
* {operations}/s
* {ops}/s
* {packets}
* {pages}/s
* {partitions}
* {processes}
* {properties}/s
* {read_repair}
* {records}
* {replicas}
* {requests}
* {requests}/s
* {responses}
* {restarts}/h
* {scanned}/{returned}
* {searches}/s
* {services}
* {sessions}
* {shards}
* {shrinks}
* {status}
* {suboperations}/s
* {syncs}
* {tasks}
* {threads}
* {tickets}
* {transactions}
* {transactions}/s
* {values}/s
* {views}
* {watches}
* {writes}/s
* {znodes}

I guess that means we need to change the mechanism a little bit to properly handle all these cases.

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [reply exporter/prometheusexporter/metric_rename.go:52]: > metrics names are typically structured as <namespace>.<object>.<metric>[_<unit>][_total]

Is that typical outside of Prometheus?  Thats what i'm trying to figure out.  If it is a prometheus metric, it is typically structured as <namespace>_<name>_<unit>_<total>, but we expect the prometheus receiver to handle this case by trimming it down to <namespace>_<name> before it gets to the exporter.

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [inline exporter/prometheusexporter/metric_rename.go:73]: If we want to handle `/`, I think we need to split by `/`, map each side to a unit (using the same strategy as above), and joining the results with _per_.  So `GiBy/h` would become `disk_write_gibibyte_per_hour` (assuming we handle GiBy and h in the map).  Right now, I think we end up with `disk_write_GiBy_per_h`, which isn't great.

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [inline exporter/prometheusexporter/metric_rename.go:78]: Can you give an example of what you are thinking with this?  I'm not sure when we expect to see other characters.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:52]: Really I'm not comfortable in making **prometheusexporter** rely strongly on the behavior or **prometheusreceiver**. If the behavior of the receiver changes in the future, it will cause issues in users' environments that will go unnoticed in our unit and integration tests.

Also, there is the case where the flow of data is slightly different. If the metrics take a little detour through Splunk/SignalFX, for example: Prometheus exporter --> SignalFX Agent --> **signalfxreceiver** --> **prometheusexporter**.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:74]: Yeah I agree, and then do the splitting as you suggested below. I'll update the PR according to your suggestions. Thank you!

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:78]: We are sanitizing metric names before exporting to Prometheus, so we must sanitize the unit before appending to the metric name. Units are not guaranteed to contain only alphanumeric chars.

Example: `°C`

Comment (dashpole) [reply exporter/prometheusexporter/metric_rename.go:78]: sg

**Comment by bertysentry** [comment]
@dashpole Complete rewrite to properly handle all metrics (and their units) in current Otel receivers. Added many unit tests to cover these cases.

Designed to be fast and robust, while covering most reasonable cases (and 100% of current Otel receivers).

**Comment by bertysentry** [comment]
@Aneurysm9 @pmm-sumo Please review! 😊

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [inline exporter/prometheusexporter/metric_rename.go:118]: would it be safer to trim total at the top from the end, and add it back here?  Many metrics are named poorly and may have total elsewhere in the name (e.g. this makes `total_ice_cream_cones_eaten -> ice_cream_cones_eaten_total`).  You might be able to keep it clean like this using `defer`:

```golang
        if cleanUnit == "1" && metric.DataType() == pmetric.MetricDataTypeGauge {
                 nameElements = removeLastItem(nameElements, "ratio")
                 defer func() {nameElements = append(nameElements, "ratio")}()
	}
```

You could also change the `contains` function, to `endsWith`, since other suffixes are trimmed already.

File: exporter/prometheusexporter/metric_rename.go
Comment (dashpole) [inline exporter/prometheusexporter/metric_rename.go:130]: We expect `{xyz}` to be the entire word, right?  E.g. `{bananas}/s`.  Could we just check to see if any part of `unitElements` contains `{` or `}`, and map it to empty-string if that is the case?

File: exporter/prometheusexporter/README.md
Comment (bertysentry) [reply exporter/prometheusexporter/README.md:58]: Many Otel receivers improperly use unit `1` for counting objects, while it's reserved for fractions and ratios.

To identify these misuses of the `1` as a unit and to avoid appending `_ratio` where it's not expected, we check that the metric is a gauge, because ratios and fractions are always gauges.

In the meantime, I also created many issues to fix these improper units in receivers.

Comment (dashpole) [reply exporter/prometheusexporter/README.md:58]: Could ratios be counters (e.g. fraction_of_work_done_ratio)?  I'm OK with it as a temporary restriction for practicality, but I'd prefer to not keep it around longer-term, as we shouldn't prevent things allowed by the OpenMetrics spec.  If there is an umbrella tracking issue for correcting the use of `1`, it would be helpful to link that by the "Append _ratio for ratio Gauges" comment.

File: exporter/prometheusexporter/metric_rename.go
Comment (bertysentry) [reply exporter/prometheusexporter/metric_rename.go:130]: I guess we can simplify here. Thanks!

File: exporter/prometheusexporter/README.md
Comment (bertysentry) [reply exporter/prometheusexporter/README.md:58]: There is no such thing as metrics suffixed with `_ratio_total` in Prometheus. It's either `_ratio` or `_total`, it really can't be both. Same for OpenMetrics. I think we're safe here ;-)

**Comment by bertysentry** [comment]
Code updated following suggestions, and rebased to solve conflict.

File: exporter/prometheusexporter/README.md
Comment (bertysentry) [reply exporter/prometheusexporter/README.md:62]: You're actually right: theoretically, it's possible.

However, `_total` indicates a monotonic counter, which keeps increasing. `_ratio` indicates a value comprised between 0 and 1. This means the metric is a function with a limit of 1 as it approaches infinite. It also means that its derivative has a limit of zero when it approaches infinite. This means that the `rate()` of this metric will be always zero after a certain time. Which is absurd.

Not appending `_ratio` to such metrics will not be a problem to any "real life" scenario, and will allow us to cover many cases where the unit was improperly specified as `1`.

I'm adding a note to the code as you suggested.

**Comment by bertysentry** [comment]
@dashpole @pmm-sumo @Aneurysm9 If you got some time, this PR needs approvals 😉

**Review by dashpole — APPROVED** [review]

**Comment by bertysentry** [comment]
@Aneurysm9 @pmm-sumo I need this PR to be merged so I can start working on **prometheusremotewriteexporter** to implement the same mechanism. Is there anyone else you would like to review these changes? Thank you!

**Comment by bertysentry** [comment]
Rebased on **main** to solve recently introduced conflict in CHANGELOG.md

**Comment by bertysentry** [comment]
Restructured the code so that common code for normalizing metric names and labels is now in module **/pkg/translator/prometheus**. Both **prometheusexporter** and **prometheusremotewriteexporter** now leverage these functions, so the behavior is consistent.

@dashpole Your review and comments are appreciated 😊
@Aneurysm9 and @pmm-sumo This PR changes a few things in the Prometheus exporters. Your comments are welcome.

I will look into the Prometheus compliance tests that are now failing.

**Review by dashpole — APPROVED** [review]

**Comment by bertysentry** [comment]
@dmitryax @alolita @Aneurysm9 This metric normalization is required as per [OpenTelemetry's official specification](https://opentelemetry.io/docs/reference/specification/metrics/datamodel/#metric-metadata-1). Please review (and merge if okay). I'm concerned these changes could cause conflicts with other PRs that update the Prometheus-specific code.

If review is not possible at this time (which I totally understand), could you please redirect me to other potential reviewers? Thank you!

**Comment by bertysentry** [comment]
I am sorry to insist, but this PR fixes an important issue in the Prometheus exporters. How can I help to get this merged?

**Comment by bertysentry** [comment]
I'm begging here: I need a review of this PR, as I can't keep resolving conflicts with upstream.
Please remember that changes in this PR are behind a feature gate, so this PR is not going to cause havoc on the project, but it refactors some duplicate code between the exporters for Prometheus (hence the many changes in this PR).
Please @jpkrohling @Aneurysm9 your help is much appreciated. Thanks!

**Comment by djaglowski** [comment]
@bertysentry, CI failed on an unrelated flaky test (#11451). I've restarted the workflow.

**Comment by bertysentry** [comment]
Daily burden of `go.mod` conflict resolution... 
image

Can anyone give me a hint on how to get these changes approved and merged? It's critical to change the behavior of the Prometheus exporters as soon as possible so they follow Otel's specifications and Prometheus' conventions. The longer we wait, the worse it gets to end users. 
Feel free to contact me on CNCF's Slack if you want to discuss the changes. @Aneurysm9 @pmm-sumo @jpkrohling 
Thank you!

**Review by djaglowski — APPROVED** [review]

**Comment by bertysentry** [comment]
Thank you @djaglowski for your help, much appreciated!

**Comment by djaglowski** [comment]
Glad to help @bertysentry.

I should have noticed this before, but since this PR introduced a new module, we really ought to have a code owner for it. Would you, @bertysentry and/or @dashpole, be willing to be the code owner for `pkg/translator/prometheus`?

**Comment by dashpole** [comment]
I'm happy to be a codeowner

**Comment by bertysentry** [comment]
Happy to!

**Comment by djaglowski** [comment]
Thank you both! I'll make the PR

**Pull request merged**
